core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -460,7 +460,9 @@ class Partition(val topic: String,
     }.map(_.logEndOffset)
     val newHighWatermark = allLogEndOffsets.min(new LogOffsetMetadata.OffsetOrdering)
     val oldHighWatermark = leaderReplica.highWatermark
-    if (oldHighWatermark.messageOffset < newHighWatermark.messageOffset || oldHighWatermark.onOlderSegment(newHighWatermark)) {
+    if (oldHighWatermark.messageOffset < newHighWatermark.messageOffset ||
+       // Ensure that watermark offset does not increase even if segmentBaseOffset of the newer LogOffsetMetadata is larger
+       (oldHighWatermark.messageOffset == newHighWatermark.messageOffset && oldHighWatermark.onOlderSegment(newHighWatermark))) {
```

Comment:
I think I'd describe it as:

"when we initially roll the segment, the offset doesn't change, but we want the high watermark to point to the latest segment"

@hachikuji, does this seem OK to you?